[pvr] Improve PVR startup time

Jalle19 opened the pull request:
People have been complaining that starting the PVR manager takes too long. The reason for this is partly that we've done some unnecessary persisting of channels, groups and group members during every startup. With these changes, not a single group or channel is being persisted during startup.

Walkthrough:
- In `CPVRChannel::UpdateFromClient`, we check if the client channel name has changed. This check has always evaluated to true because we've never stored that field in the database. This means that we've always persisted every channel at least once on every startup. Thanks @FernetMenta for spotting it.
- I can't remember the reason for 98d4a4c
- we've always loaded group members into a temporary group, which means the code has always thought that every group we've loaded from the database has been empty, even when that's not the case. @opdenkamp do you know why this has been done like this? `git blame` doesn't go far enough.
- when loading groups from the database we've created an empty group first, then updated it with the data from the database. This update has caused `m_bChanged` to be set to true for every group loaded from the database, which has lead to _surprise_ all groups being persisted once more during startup.
- we've added every EPG tag to `m_changedTags` on startup. This vector is then looped over and tags are persisted if they've been changed. We can make the loop shorter by only considering tags that have actually changed. The gain here is most likely negligible on most devices, but still.

I've only tested this in VS so far on a reasonably fast machine and I can't even see the "Loading channels from clients" anymore. Next up is to have a look through the EPG loading, I'm sure there's something that can be optimized there as well.

@xhaggi pinging you too

Jalle19 committed:
[pvr] copy some group properties immediately when the group is loaded
from the database, otherwise the new group will be marked as "changed"
a bit further down which causes all groups to be persisted every time
the PVR manager starts

Jalle19 commented:
If anyone who sees this has hundreds or even thousands of channels, please try this and leave a comment if it improves anything.

hudokkow commented:
I've noticed this behavior also and found it odd.
For example, I'm currently developing a new PVR addon and noticed a few things:

1. I push a change and recompile my addon (~2m) without changing anything in settings.xml or the backend
2. Start Kodi fresh with a new PVR addon copy
3. PVR loads channels again but channels, channel group(s) or members didn't changed
4. Loads EPG even thought I have a few days of data in the DB and again, nothing changed in the last 2m, not significantly atleast
5. Load recordings/scheduled recordings, if available and scheduled

On my test setup using 3 channel groups with 300 to 400 channels (worst case scenario and certainly not my production setup) it takes around 1m to fully load PVR although it's usable before that time.

The point is PVR shouldn't be loading everything again and again. If nothing changed, atleast not significantly, it should defer loading/parsing/updating and start PVR section as soon as possible.
Channel group name didn't change? Trust it and work with the channels already in the DB. Chances are most users don't change channels or group name often.

Same for EPG. If we already have 3 days of (valid) data in the DB, trust it.

Recordings and timers are a bit trickier. Those tend to change quite often, specially if you have more than one Kodi instance running, like I do.

For the record, this isn't a problem for me. I tend to leave my hardware (mostly RPis) running 24/7 and Kodi does a good job keeping everything synced with the backend.

On another note, if Jenkins isn't busy with more important PRs, trigger an Ubuntu build and I'll happily test. Saves me a build tonight. ;)

Jalle19 commented:
I don't know how to do that, sorry.

MilhouseVH commented:
Presumably this will be of benefit to Pi users?  I don't use PVR myself, so I've no idea how slow it might be right now. I'll hopefully include this PR in tonight's OpenELEC/Pi test build.

Memphiz commented:
fyi - our jenkins can't build for ubuntu ;)

hudokkow commented:
MilhouseVH, even better. Thanks.
Memphiz, lol.

opdenkamp commented:
will need to take a look when i have time, but if it's slow to start, then it sounds like a regression in helix that needs to be fixed. it wasn't slow before, and my own setup has hundreds of channels, with several days of epg data for each channel, and most of them are running on an nfs root or an (internal) sd card. importing all of this takes 2-3 seconds when it's a new installation, and about 1 second when it's an existing installation.

Jalle19 commented:
@MilhouseVH it should benefit everyone, but especially Raspberry Pi and other slow devices. Would be great if you can include it in any of your test builds since I don't have a Pi with Kodi installed at the moment.

@opdenkamp I haven't noticed any performance regressions, this is just an improvement (not a regression fix). The majority of the startup time is now spent loading the EPG, but I checked and at least that slowness is not due to unnecessary database writes during startup so changes to that part is for another PR.

xhaggi commented:
IMO looks sane and +1

opdenkamp commented:
there were several issues with this when i took a quick look, will 
comment this week end.

On 23-01-15 17:08, Sascha Woo wrote:
>
> IMO looks sane and +1
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/xbmc/xbmc/pull/6246#issuecomment-71216260>.
>

MilhouseVH commented:
Initial feedback here...

http://forum.kodi.tv/showthread.php?tid=211501&pid=1901604#pid1901604

Jalle19 commented:
@MilhouseVH very strange, I never had that issue when testing. Can someone else try this?

opdenkamp commented:
didn't have time to review any PR because of the stupid flu...

FernetMenta commented:
@Jalle19 I can confirm the issue reported by MilhouseVH. Right after startup all backend channels numbers are 0. They seem to get correct values after having cycled once through all groups.

FernetMenta commented:
why are backend channel numbers persisted and compared on updateFromClient? does not make much sense to me. A reorder on the backend would slow down startup again.

Jalle19 commented:
IMO what we need is to store the backend channel number in a separate column in the database, that way we could partially avoid all this wizardry with "PreventSortAndRenumber" etc. We do the same with the channel name (i.e. we store both).

Jalle19 commented:
@FernetMenta I still can't reproduce the issue (maybe I've made my database immune to this while working on this PR), neither with nor without backend channel numbers enabled.

FernetMenta commented:
@Jalle19 @opdenkamp I don't know what m_iCachedChannelNumber is good for but it is zero when first accessed by the skin. Note that iChannelNumber holds the correct value.

opdenkamp commented:
that's an expected result from these changes, which is one of the issues 
that i saw when i checked it quickly. iChannelNumber != channel number 
in each group, except when you have backend channel numbers enabled.

so don't merge this in before i had a chance to review it properly. 
stupid flu virus prevented me from taking a look at the open PRs, sorry.

On 26-01-15 12:51, Rainer Hochecker wrote:
>
> @Jalle19 <https://github.com/Jalle19> @opdenkamp 
> <https://github.com/opdenkamp> I don't know what 
> m_iCachedChannelNumber is good for but it is zero when first accessed 
> by the skin. Note that iChannelNumber holds the correct value.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/xbmc/xbmc/pull/6246#issuecomment-71449867>.
>

Jalle19 commented:
Just tried again by removing TV28.db (the new version) and using TV27.db which was previously created when on the master branch, and still I wasn't unable to reproduce this issue you're talking about. Which exact settings do you have enabled regarding channel numbers (order and numbering)?

Jalle19 commented:
Can someone that uses tvheadend test this as well? So far only MythTV and VDR users have tested this, maybe there's something peculiar about the way the addons do the initial startup?

opdenkamp commented:
should have a bit of time to review open PRs in the evening today.

FernetMenta commented:
@Jalle19 I have "always use the channel order ...." and "Use backend channel ..." enabled.

hudokkow commented:
@Jalle19, been testing this with Vuplus PVR on my RPIs and can't reproduce/see any problem.
BTW, this PR consistently gives me a 10% improvement in loading time with 126 channels under one group (from over a minute to ~53s).

opdenkamp commented:
1 minute?? are you looking at importing channels or the epg?
anything over 2 seconds = broken, even on a slow ass rpi.

On 28-01-15 14:46, h.udo wrote:
>
> @Jalle19 <https://github.com/Jalle19>, been testing this with Vuplus 
> PVR on my RPIs and can't reproduce/see any problem.
> BTW, this PR consistently gives me a 10% improvement in loading time 
> with 126 channels under one group (from over a minute to ~53s).
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/xbmc/xbmc/pull/6246#issuecomment-71836593>.
>

hudokkow commented:
Those values refer to:
- Boot RPi 
- Kodi's UI starts
- Message "Loading channels bla-bla-bla..." shown (start chronograph)
- Load channels, EPG, recordings, timers, etc... (stop chronograph)

Result is the above. Like I said in an earlier comment (see above), PVR is usable before this process ends, usually after a couple of seconds.

dr88dr88 commented:
Sins Kodi release EPG loading is extreme slow in the VuPlus add-on. so I agree that it is broken.

opdenkamp commented:
without seeing logs, I bet it's not using the same unique IDs all the time for channels

Jalle19 commented:
@opdenkamp I changed the first commit to just ignore the client channel name, removed the two after that (cosmetics and internal group handling thing). I'm still not convinced that the LoadFromClients for each group takes so long that we have to create a set of hacks just because, it seems to take a split second.

opdenkamp commented:
it's not about taking long, it's about first importing each channel and having their db IDs, channel numbers and other properties set, which are then used by group members.

FernetMenta commented:
can we revert #6091 until this is fixed properly? waiting a minute on startup is not acceptable during development.

Jalle19 commented:
@FernetMenta I still don't see how that patch makes anything slower.

FernetMenta commented:
I told you already. It persists every single channel when updating from backend.

MilhouseVH commented:
It's not negative feedback, but it's hardly positive either: [forum post](http://forum.kodi.tv/showthread.php?tid=211501&pid=1914177#pid1914177).

I'll drop this PR from my builds as I'm not sure it's really doing what it's supposed to do.

popcornmix commented:
Another [forum post](http://forum.kodi.tv/showthread.php?tid=211501&pid=1918625#pid1918625) reports a 3 second speed-up (now takes 22 seconds), so there is a benefit (although still pretty slow).

da-anda commented:
the mentioned 3sec speed up was tested by me. I'm running TVH on the Wetek box and remote connected my PI2. With this patch in the boot time from rainbow image until EPG import is finished was up to 3 seconds faster (did several reboots with and without this PR and got an average of 2sec speed improvement) and that with only ~60 channels on my end. So in my tests it at least made a difference.

fritsch commented:
Do you do those measurements with: Do not store the EPG in the database
disabled?

2015-02-08 16:33 GMT+01:00 da-anda <notifications@github.com>:

> the mentioned 3sec speed up was tested by me. I'm running TVH on the Wetek
> box and remote connected my PI2. With this patch in the boot time from
> rainbow image until EPG import is finished was up to 3 seconds faster (did
> several reboots with and without this PR and got an average of 2sec speed
> improvement) and that with only ~60 channels on my end. So in my tests it
> at least made a difference.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/6246#issuecomment-73415722>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

da-anda commented:
this option is enabled - so no EPG caching inside Kodi and always importing from backend (which is supposed to be faster then first loading from Kodi db and then updating from backend). I can compare with this option disabled as well.

fritsch commented:
Nope, that's correct. Keep it "enabled", disabling (which is the insane default) will produce more load to the disk.

opdenkamp commented:
it's the not-so-insane default because there's hardware with an integrated backend that doesn't have the memory to provide the full epg :)
but indeed, it'll lead to more disk load, which makes a huge difference on a pi or anything else using a slow disk.

we could add some flag to the add-on xml that allows add-ons to tell kodi that it needs kodi's epg db and disable it by default if it's not set.

da-anda commented:
we should move the talk about the EPG import over to the forum.

So is this PR good to be merged or are there any objections? If time allows I could do another test over at a friends place who complained that the PVR startup using MediaPortal as backend is taking ages for him

opdenkamp commented:
EPG import is not important for "startup speed". If you remove the progress bar, then people won't even notice that it's still loading :)

0152a992fb7d9adbeec2adb4e924cfa07312e2ee is wrong as explained before and 0be84113b8c30b722d120e9779b1ab15bec05f89 was done differently by @FernetMenta already.

Jalle19 commented:
I've removed the first two commits that were the main cause for discussion. Can you guys take a look at what's remaining?

xhaggi commented:
fine by me

Jalle19 commented:
@opdenkamp @FernetMenta what do you think?

opdenkamp commented:
+1

Jalle19 commented:
jenkins build this please

FernetMenta commented:
looks ok to me

Jalle19 commented:
jenkins build this please

Jalle19 commented:
jenkins build this please
